types/int_test.go

```diff
@@ -109,3 +110,342 @@ func TestUint(t *testing.T) {
+			{i1.LT(i2), n1 < n2},
+		}
+
+		for _, tc := range cases {
+			require.Equal(t, tc.nres, tc.ires)
+		}
+	}
+}
+
+func randint() Int {
+	return NewInt(rand.Int63())
+}
+
+func TestImmutabilityInt(t *testing.T) {
+	ops := []func(*Int){
+		func(i *Int) { _ = i.Add(randint()) },
+		func(i *Int) { _ = i.Sub(randint()) },
+		func(i *Int) { _ = i.Mul(randint()) },
+		func(i *Int) { _ = i.Div(randint()) },
+		func(i *Int) { _ = i.AddRaw(rand.Int63()) },
+		func(i *Int) { _ = i.SubRaw(rand.Int63()) },
+		func(i *Int) { _ = i.MulRaw(rand.Int63()) },
+		func(i *Int) { _ = i.DivRaw(rand.Int63()) },
+		func(i *Int) { _ = i.Neg() },
+		func(i *Int) { _ = i.IsZero() },
+		func(i *Int) { _ = i.Sign() },
+		func(i *Int) { _ = i.Equal(randint()) },
+		func(i *Int) { _ = i.GT(randint()) },
+		func(i *Int) { _ = i.LT(randint()) },
+		func(i *Int) { _ = i.String() },
+	}
+
+	for i := 0; i < 20; i++ {
+		n := rand.Int63()
+		ni := NewInt(n)
+
+		for _, op := range ops {
+			op(&ni)
+
+			require.Equal(t, n, ni.Int64())
+			require.Equal(t, NewInt(n), ni)
+		}
+	}
+}
+
+func TestImmutabilityUint(t *testing.T) {
+	ops := []func(*Uint){
+		func(i *Uint) { _ = i.Add(randuint()) },
+		func(i *Uint) { _ = i.Sub(NewUint(rand.Uint64() % i.Uint64())) },
+		func(i *Uint) { _ = i.Mul(randuint()) },
+		func(i *Uint) { _ = i.Div(randuint()) },
+		func(i *Uint) { _ = i.AddRaw(rand.Uint64()) },
+		func(i *Uint) { _ = i.SubRaw(rand.Uint64() % i.Uint64()) },
+		func(i *Uint) { _ = i.MulRaw(rand.Uint64()) },
+		func(i *Uint) { _ = i.DivRaw(rand.Uint64()) },
+		func(i *Uint) { _ = i.IsZero() },
+		func(i *Uint) { _ = i.Sign() },
+		func(i *Uint) { _ = i.Equal(randuint()) },
+		func(i *Uint) { _ = i.GT(randuint()) },
+		func(i *Uint) { _ = i.LT(randuint()) },
+		func(i *Uint) { _ = i.String() },
+	}
+
+	for i := 0; i < 20; i++ {
+		n := rand.Uint64()
+		ni := NewUint(n)
+
+		for _, op := range ops {
+			op(&ni)
+
+			require.Equal(t, n, ni.Uint64())
+			require.Equal(t, NewUint(n), ni)
+		}
+	}
+
+}
+
+func randuint() Uint {
+	return NewUint(rand.Uint64())
+}
+
+func TestEncodingRandom(t *testing.T) {
+	for i := 0; i < 20; i++ {
+		n := rand.Int63()
+		ni := NewInt(n)
+		var ri Int
+
+		str, err := ni.MarshalAmino()
+		require.Nil(t, err)
+		err = (&ri).UnmarshalAmino(str)
+		require.Nil(t, err)
+
+		require.Equal(t, ni, ri)
+
+		bz, err := ni.MarshalJSON()
+		require.Nil(t, err)
+		err = (&ri).UnmarshalJSON(bz)
+		require.Nil(t, err)
+
+		require.Equal(t, ni, ri)
+	}
+
+	for i := 0; i < 20; i++ {
+		n := rand.Uint64()
+		ni := NewUint(n)
+		var ri Uint
+
+		str, err := ni.MarshalAmino()
+		require.Nil(t, err)
+		err = (&ri).UnmarshalAmino(str)
+		require.Nil(t, err)
+
+		require.Equal(t, ni, ri)
+
+		bz, err := ni.MarshalJSON()
+		require.Nil(t, err)
+		err = (&ri).UnmarshalJSON(bz)
+		require.Nil(t, err)
+
+		require.Equal(t, ni, ri)
+	}
+}
+
+func TestEncodingTableInt(t *testing.T) {
+	var i Int
+
+	cases := []struct {
+		i   Int
+		str []byte
+	}{
+		{NewInt(0), []byte("\"0\"")},
```

Comment:
Awesome! Thanks for making some regression tests! Can you add MarshalAmino as another field in here?